src/content/blog/2026-12-08-web-monetization-open-payments-part-2-connecting-wallet.md

```diff
@@ -0,0 +1,122 @@
+
+```ts
+function documentMutationCallback() {
+  const linkElements = getLinkElements()
+  const addedLinkElements = diff(linkElements, existingLinkElements)
+  const removedLinkElements = diff(existingLinkElements, linkElements)
+
+  addedLinkElements.forEach((elem) => {
+    const id = assignId(elem)
+    const walletAddressInfo = await validateAndFetch(elem.href)
+    dispatch('start_monetization', { walletAddressInfo, id })
+  })
+
+  removedLinkElements.forEach((elem) => {
+    const id = getAssignedId(elem)
+    dispatch('stop_monetization', { id })
+  })
+}
+```
+
+All of these state changes — discovery, modification, and removal — are communicated via message passing to the background script. This ensures the background script, which manages the actual payment logic, is always perfectly synchronized with the live state of the web page. This synchronization ultimately feeds the pop-up to display accurate, real-time monetization status to the user.
+
+## Payment Session
+
+A Payment Session in the background script acts as the operational counterpart for a monetization link element. It uses the unique ID from the content script to maintain a link between the element in the DOM and its background representation. When a new payment session is initiated, it must undergo a series of checks.
+
+First, it verifies whether the user's connected wallet address is authorized to send payments to the specified wallet address. The extension sends a “[create quote](https://openpayments.dev/apis/resource-server/operations/create-quote/)” request to the receiver’s resource server endpoint. If this request results in an "invalid receiver" error, it means the two wallets are not "peered" or connected. This may occur due to legal or technical reasons. For example, a wallet on the test namespace (e.g., [Interledger Test Wallet](https://wallet.interledger-test.dev/)) _fake money_ cannot send a payment involving real money to a production wallet (e.g. [Interledger Wallet](https://interledger.app/) or GateHub). Consequently, we mark this payment session as "invalid", and no payments can proceed during the browsing session for this wallet address.
+
+The Payment Session object also has some helper methods to manage the payments effectively: it includes a method to send a specified amount of money, a method for dispatching `MonetizationEvent` to corresponding link elements (next article), and helpers to check if wallet address is disabled, paused or payable, ensuring the recipient's status is constantly confirmed to prevent failed payments or unnecessary transactions.
```

Comment:
```suggestion
The Payment Session object also has some helper methods to manage the payments effectively: it includes a method to send a specified amount of money, a method for dispatching `MonetizationEvent` to corresponding link elements (next article), and helpers to check if a wallet address is disabled, paused or payable, ensuring the recipient's status is constantly confirmed to prevent failed payments or unnecessary transactions.
```